policy: Report reason inputs are nonstandard from AreInputsStandard

Empact opened the pull request:
This results in a more informative debug string to the validation state in the
case of failure.

promag commented on src/policy/policy.cpp:
```diff
@@ -169,19 +170,28 @@ bool AreInputsStandard(const CTransaction& tx, const CCoinsViewCache& mapInputs)
-            if (stack.empty())
+            }
+            if (stack.empty()) {
+                reason = strprintf("input %u script stack empty", i);
```
Not tested?

promag commented on src/policy/policy.h:
```diff
@@ -88,9 +88,10 @@ bool IsStandardTx(const CTransaction& tx, std::string& reason);
     /**
      * Check for standard transaction types
      * @param[in] mapInputs    Map of previous transactions that have outputs we're spending
+     * @param[out] reason      If return is false, contains the reason the inputs were judged non-standard
```
`contains the reason of the first non-standard input`?

promag commented on src/policy/policy.cpp:
```diff
@@ -169,19 +170,28 @@ bool AreInputsStandard(const CTransaction& tx, const CCoinsViewCache& mapInputs)
-            if (!EvalScript(stack, tx.vin[i].scriptSig, SCRIPT_VERIFY_NONE, BaseSignatureChecker(), SigVersion::BASE))
+            ScriptError serror;
+            if (!EvalScript(stack, tx.vin[i].scriptSig, SCRIPT_VERIFY_NONE, BaseSignatureChecker(), SigVersion::BASE, &serror)) {
+                reason = strprintf("input %u script error: %s", i, ScriptErrorString(serror));
```
Not tested?

promag commented on src/bench/ccoins_caching.cpp:
```diff
@@ -77,8 +77,10 @@ static void CCoinsCaching(benchmark::State& state)
-        assert(success);
+        std::string reason;
+        if (!AreInputsStandard(t1, coins, reason)) {
+            throw std::runtime_error(reason);
```
Why this change? It must succeed no?

promag reviewed (commented):
Concept ACK.

Empact committed:
Propagate validation debug information from AreInputsStandard

This adds specific debug information including the input index and manner of
failure.

Empact commented:
@promag Addressed your comments, and after a fair amount of banging my head against the wall, managed to test all cases.

sipa commented:
utACK 7298231aa8faaff65c261eac557448c223b76500

luke-jr commented:
This is already part of #7533 

At the very least, please try to use the short rejection reasons therein...

luke-jr commented:
https://github.com/bitcoin/bitcoin/pull/7533/commits/c59e7166253e962043b5fdccd5de8a21969d21e0

Empact commented:
Thanks @luke-jr, I adapted the short reasons from c59e716. How does that look? Is it alright to put the input index in the rejection reason or better to append?

Empact commented:
Also moved the string instantiation out of the bench loop.

luke-jr commented:
Might be better to have the index at the end, not sure.

Empact commented:
@luke-jr moved the input index to the end. Now I'm questioning the need for the `bad-txns-` prefix, given that these are to become debug output, rather than the reason string, which is what usually bears this prefix.

luke-jr commented:
Why do you want to limit it to debug output? Reasons should be in the reason field.

Empact commented:
@luke-jr Good point - split the reason and debug fields, and replaced `bad-txns-nonstandard-inputs` with one of: `bad-txns-script-nonstandard`, `bad-txns-scriptsig-error`, or `bad-txns-scriptcheck-sigops`.

luke-jr commented:
Prefer to use the existing `bad-txns-input-script-unknown`, `bad-txns-input-scriptsig-failure`, `bad-txns-input-scriptcheck-missing`, and `bad-txns-input-scriptcheck-sigops`

Empact commented:
Rebased for  #13429

Empact commented:
Fixed rebase for #13429, dropped refactor commit, and made use of every reason recommended by @luke-jr

practicalswift commented on src/test/script_p2sh_tests.cpp:
```diff
@@ -362,8 +378,47 @@ BOOST_AUTO_TEST_CASE(AreInputsStandard)
+    txToNonStd4.vin.resize(1);
+    txToNonStd4.vin[0].prevout.n = 8;
+    txToNonStd4.vin[0].prevout.hash = txFrom.GetHash();
+    txToNonStd4.vin[0].scriptSig = CScript(&op_return[0], &op_return[sizeof(op_return)]);
```
Isn't `&op_return[sizeof(op_return)]` an array overrun?

practicalswift commented on src/test/script_tests.cpp:
```diff
@@ -1032,6 +1032,16 @@ BOOST_AUTO_TEST_CASE(script_PushData)
+
+    ScriptError err;
+    std::vector<std::vector<unsigned char>> op_return_stack;
+    BOOST_CHECK(!EvalScript(op_return_stack, CScript(&op_return[0], &op_return[sizeof(op_return)]), SCRIPT_VERIFY_P2SH, BaseSignatureChecker(), SigVersion::BASE, &err));
```
Same here?

Empact commented:
@practicalswift replaced uses of `&array[sizeof(array]` with `array + sizeof(array)`

Empact commented:
Stopped shadowing `reason` var (by reusing the existing var).
Limited test array access changes to only affecting the added tests.

Empact commented:
Added code-blocks to tests to protect against accidental reuse, e.g. of `txToNonStd1` in other test cases.

sipa commented:
utACK 85e2ddb14c15b9613093ec0cb991b77091c4287c, with or without nits addressed.

Empact commented:
Updated:
* `bad-txns-input-scriptsig-failure` -> `bad-txns-input-p2sh-scriptsig-malformed`
* `bad-txns-input-scriptcheck-missing` -> `bad-txns-input-p2sh-no-redeemscript`
* `bad-txns-input-scriptcheck-sigops` -> `bad-txns-input-p2sh-redeemscript-sigops`

luke-jr commented on src/policy/policy.cpp:
```diff
@@ -169,16 +169,28 @@ bool AreInputsStandard(const CTransaction& tx, const CCoinsViewCache& mapInputs)
-            if (!EvalScript(stack, tx.vin[i].scriptSig, SCRIPT_VERIFY_NONE, BaseSignatureChecker(), SigVersion::BASE))
+            ScriptError serror;
+            if (!EvalScript(stack, tx.vin[i].scriptSig, SCRIPT_VERIFY_NONE, BaseSignatureChecker(), SigVersion::BASE, &serror)) {
+                reason = "bad-txns-input-scriptsig-failure";
```
The order of words typically indicates subcategories/relationships. For that reason, "bad-txns-scriptsig-p2sh-malformed" would fit better.

Empact replied:
I’m currently using: “bad-txns-input-p2sh-scriptsig-malformed” for this.

Because all reasons in this function relate to inputs, of which are tested/subdivided according to type (p2sh in this case), then the subject scrptsig, then the issue.

Not that there’s a right or wrong here, but that’s the rationale.

MarcoFalke commented on src/validation.cpp:
```diff
@@ -706,8 +706,9 @@ static bool AcceptToMemoryPoolWorker(const CChainParams& chainparams, CTxMemPool
-        if (fRequireStandard && !AreInputsStandard(tx, view))
-            return state.Invalid(ValidationInvalidReason::TX_NOT_STANDARD, false, REJECT_NONSTANDARD, "bad-txns-nonstandard-inputs");
+        std::string debug;
+        if (fRequireStandard && !AreInputsStandard(tx, view, reason, debug))
```
style-nit: Could add `{`?

Empact replied:
Added 👍

maflcko reviewed (commented):
Concept ACK

Empact commented:
Rebased, added missing braces.

Empact commented:
Rebased

sipa commented:
utACK 6a241604fec4346d90495fd468f54a6bf6a0ee88

Empact commented:
Rebased for #18859

0xB10C commented:
Concept ACK

glozow reviewed (commented):
Concept ACK to more informative error messages

Empact committed:
refactor: Convert AreInputsStandard to return results via std::opt<std::pair>

theStack commented:
Concept ACK

theStack commented on src/policy/policy.cpp:
```diff
@@ -186,20 +187,36 @@ bool AreInputsStandard(const CTransaction& tx, const CCoinsViewCache& mapInputs)
             // validation.
+            reason = "bad-txns-input-witness-unknown";
+            debug = strprintf("input %u", i);
             return false;
```
This if-branch is currently dead code, since the type `WITNESS_UNKNOWN` is already matched in the first if-condition a few lines above (obvioulsy, it was forgotten to be removed there). I.e. we want

```diff
diff --git a/src/policy/policy.cpp b/src/policy/policy.cpp
index 08f41376a..581fdcce3 100644
--- a/src/policy/policy.cpp
+++ b/src/policy/policy.cpp
@@ -186,7 +186,7 @@ std::optional<std::pair<std::string, std::string>> AreInputsStandard(const CTran

         std::vector<std::vector<unsigned char> > vSolutions;
         TxoutType whichType = Solver(prev.scriptPubKey, vSolutions);
-        if (whichType == TxoutType::NONSTANDARD || whichType == TxoutType::WITNESS_UNKNOWN) {
+        if (whichType == TxoutType::NONSTANDARD) {
             return std::make_pair(
                 "bad-txns-input-script-nonstandard",
                 strprintf("input %u", i)

```

To avoid issues like this, should probably also add a test for "bad-txns-input-witness-unknown"?

Empact replied:
Good point! Updated here: https://github.com/bitcoin/bitcoin/compare/631d3c980090343f2f0d9dd97dac585c7fa94f13..1f3c017cce798bdcee1d719bf160dc07783239f2

theStack reviewed (commented):
CI fails, seems like some functional tests need to be adapted to match the more specific failure reason string (https://cirrus-ci.com/task/6048174287618048?logs=ci#L3337):
```
...
AssertionError: [node 1] Expected messages "['bad-txns-input-script-nonstandard']" does not partially match log:
...
```

Empact commented:
> CI fails, seems like some functional tests need to be adapted to match the more specific failure reason string (https://cirrus-ci.com/task/6048174287618048?logs=ci#L3337):
> 
> ```
> ...
> AssertionError: [node 1] Expected messages "['bad-txns-input-script-nonstandard']" does not partially match log:
> ...
> ```

Fixed: https://github.com/bitcoin/bitcoin/compare/1f3c017cce798bdcee1d719bf160dc07783239f2..dfa194343fde1557420a3b4a52b6a93e42adb848

achow101 commented on src/test/script_p2sh_tests.cpp:
```diff
@@ -368,8 +390,63 @@ BOOST_AUTO_TEST_CASE(AreInputsStandard)
+    txWitnessUnknown.vin.resize(1);
+    txWitnessUnknown.vin[0].prevout.n = 9;
+    txWitnessUnknown.vin[0].prevout.hash = txFrom.GetHash();
+    txWitnessUnknown.vin[0].scriptSig << std::vector<unsigned char>(witnessUnknown.begin(), witnessUnknown.end());
```
In ba538fba899482f5549a6377d657d9d89b13bfd5 "Propagate validation debug information from AreInputsStandard"

nit: setting `scriptSig` is unnecessary as the thing being checked is the scriptPubKey of the output being spent.

achow101 reviewed (commented):
The more descriptive reason strings seem to have gotten lost in the past couple of years.

aureleoules reviewed (commented):
Instead of returning an option would the Result class make more sense?
Also, instead of a pair of strings, maybe create a new struct?

achow101 commented:
Closing this as it has not had any activity in a while. If you are interested in continuing work on this, please leave a comment so that it can be reopened.
